File: MergeSort.py
def merge_sort(arr):

    # Do we still have elements to split?
    if len(arr) > 1:

        # Yes! Split the array in half
        mid = len(arr) // 2
        leftHalf = arr[:mid]
        rightHalf = arr[mid:]

        # Recursively merge each half
        merge_sort(leftHalf)
        merge_sort(rightHalf)
        
        leftIndex = 0
        rightIndex = 0
        arrayIndex = 0

        # Sort until we've reached the end of one array half...
        while leftIndex < len(leftHalf) and j < len(rightHalf):
            
            if leftHalf[leftIndex] < rightHalf[rightIndex]:
                
                arr[arrayIndex] = leftHalf[leftIndex]              
                leftIndex += 1
                
            else:
                
                arr[arrayIndex] = rightHalf[rightIndex]              
                rightIndex += 1
                
            arrayIndex += 1

        # Append the unfinished and sorted other half w/ A or B loop

        # A
        while leftIndex < len(leftHalf):
            
            arr[arrayIndex] = leftHalf[leftIndex]
            leftIndex += 1
            arrayIndex += 1
            
        # B
        while rightIndex < len(rightHalf):
            
            arr[arrayIndex] = rightHalf[rightIndex]
            rightIndex += 1
            arrayIndex += 1

File: MergeSort/MergeSort.cs
/// ------------------------------------------------------------
/// Name: MergeSort
/// Author: Matthew Collard
/// Purpose: Construct the Merge Sort algorithm
/// ------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MergeSort
{
    class MergeSort
    {
        static void Main(string[] args)
        {
            int[] aintArray = new int[] { 3,2,4,9,13,11,32,1,28 };
            BeginMergeSort(aintArray);
            Console.WriteLine(String.Join(",", aintArray));
            Console.ReadLine();
        }

        private static void BeginMergeSort(int[] aintArrayToSort)
        {
            // Anything to sort
            if (aintArrayToSort.Count() > 1)
            {
                int intMidpoint = aintArrayToSort.Length / 2;

                int[] aintLeftHalf = aintArrayToSort.Take(intMidpoint).ToArray();
                int[] aintRightHalf = aintArrayToSort.Skip(intMidpoint).ToArray();

                // Recursively sort and merge
                BeginMergeSort(aintLeftHalf);
                BeginMergeSort(aintRightHalf);

                int intLeftIndex = 0;
                int intRightIndex = 0;
                int intArrayIndex = 0;
                
                // Sort until we've reach the end of one half
                while (intLeftIndex < aintLeftHalf.Length && intRightIndex < aintRightHalf.Length)
                {
                    if (aintLeftHalf[intLeftIndex] < aintRightHalf[intRightIndex])
                    {
                        aintArrayToSort[intArrayIndex] = aintLeftHalf[intLeftIndex];
                        intLeftIndex += 1;
                    }
                    else
                    {
                        aintArrayToSort[intArrayIndex] = aintRightHalf[intRightIndex];
                        intRightIndex += 1;
                    }

                    intArrayIndex += 1;
                }

                // Append the unfinished and sorted other half w/ A or B loop

                // A
                while (intLeftIndex < aintLeftHalf.Length)
                {
                    aintArrayToSort[intArrayIndex] = aintLeftHalf[intLeftIndex];
                    intLeftIndex += 1;
                    intArrayIndex += 1;
                }

                while (intRightIndex < aintRightHalf.Length)
                {
                    aintArrayToSort[intArrayIndex] = aintRightHalf[intRightIndex];
                    intRightIndex += 1;
                    intArrayIndex += 1;
                }
                
            }
        }
    }
}
